hw/model/src/lcc.rs

```diff
@@ -0,0 +1,152 @@
+// Licensed under the Apache-2.0 license
+//
+// Derived from OpenTitan project with original copyright:
+//
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::iter;
+use std::time::Duration;
+
+use anyhow::{bail, Context, Result};
+use thiserror::Error;
+
+use caliptra_hw_model::lcc::{LcCtrlReg, LcCtrlStatus, LcCtrlTransitionCmd};
+use caliptra_hw_model::openocd::openocd_jtag_tap::{JtagTap, OpenOcdJtagTap};
+use mcu_rom_common::{Lifecycle, LifecycleControllerState};
+use poll_common::poll_until;
+
+/// Errors related to the LCC operations below.
+#[derive(Error, Debug)]
+pub enum LccUtilError {
+    #[error("LCC is not initialized; STATUS: {0}")]
+    NotInitialized(u32),
+    #[error("LCC transition mutex is already claimed.")]
+    MutexAlreadyClaimed,
+    #[error("Failed to claim the LCC transition mutex (value = 0x{0:x}).")]
+    FailedToClaimMutex(u32),
+    #[error("Bad post transition state (state = {0}).")]
+    BadPostTransitionState(String),
+    #[error("Functionality unimplemented.")]
+    Unimplemented,
+}
+
+pub fn lc_token_to_words(bytes: &[u8]) -> [u32; 4] {
+    let mut out_words = [0u32; 4];
+    bytes
+        .chunks_exact(std::mem::size_of::<u32>())
+        .map(|chunk| u32::from_le_bytes(chunk.try_into().unwrap()))
+        .zip(&mut out_words)
+        .for_each(|(word, out)| *out = word);
+    out_words
+}
+
+/// Read the LC state.
+pub fn read_lc_state(tap: &mut OpenOcdJtagTap) -> Result<LifecycleControllerState> {
+    // Check the LCC is initialized and ready.
+    let lcc_status = tap.read_lc_ctrl_reg(&LcCtrlReg::Status)?;
+    if LcCtrlStatus::from_bits_truncate(lcc_status)
+        != LcCtrlStatus::INITIALIZED | LcCtrlStatus::READY
+    {
+        bail!(LccUtilError::NotInitialized(lcc_status));
+    }
+
+    // Read the state register.
+    let lc_state = LifecycleControllerState::from(tap.read_lc_ctrl_reg(&LcCtrlReg::LcState)?);
+
+    Ok(lc_state)
+}
+
+/// Perform an LC transition.
+pub fn lc_transition(
+    tap: &mut OpenOcdJtagTap,
+    target_lc_state: LifecycleControllerState,
+    token: Option<[u32; 4]>,
+) -> Result<LifecycleControllerState> {
+    // Check the LCC is initialized and ready.
+    let lcc_status = tap.read_lc_ctrl_reg(&LcCtrlReg::Status)?;
+    if LcCtrlStatus::from_bits_truncate(lcc_status)
+        != LcCtrlStatus::INITIALIZED | LcCtrlStatus::READY
+    {
+        bail!(LccUtilError::NotInitialized(lcc_status));
+    }
+
+    // Attempt to claim the LC transition mutex if it has not been claimed yet.
+    const MULTI_TRUE: u32 = 0x96;
```

Comment:
@swenson @zhalvorsen : are these mubi values defined somewhere globally in one of the repos like on OT: https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/base/multibits.h;l=6?q=multi&sq=&ss=opentitan%2Fopentitan ?

Reply:
Not that I'm aware of